Add auto-retry on connection failure.

bemasc opened the pull request:
Partially fixes #1825 

Looks like this:
![retrying](https://cloud.githubusercontent.com/assets/191945/9418357/696f5f68-4821-11e5-81b4-ab728837acf5.png)
![reconnect_failed](https://cloud.githubusercontent.com/assets/191945/9397665/02a48380-476d-11e5-9403-88662a0269c0.png)


@lucyhe @jpevarnek

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy/1836)
<!-- Reviewable:end -->

salomegeo commented on src/generic_ui/polymer/root.html:
```diff
@@ -375,12 +378,12 @@
         </p>
       </div>
       <paper-button class='dialogButton' on-tap='{{ revertProxySettings }}'>{{ "CONTINUE_BROWSING" | $$ }}</paper-button>
-      <paper-button class='dialogButton' on-tap='{{ restartProxying }}'>{{ 'RESTART_PROXYING' | $$ }}</paper-button>
```
maybe remove this from messages.json too?

bemasc replied:
I added this back in.

lucyhe commented:
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

<sup>**[src/generic_ui/polymer/root.html, line 383 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1836#-JxG1riZXsgAGapszMK8)** ([raw file](https://github.com/uproxy/uproxy/blob/a45ff8c47a4f67034a47432d51064cfaea35856b/src/generic_ui/polymer/root.html#L383)):</sup>
Personal preference is "Reconnecting" would look better if text were smaller

---

<sup>**[src/generic_ui/scripts/ui.ts, line 372 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1836#-JxG1cj6XgmLgQpoH1ZJ)** ([raw file](https://github.com/uproxy/uproxy/blob/a45ff8c47a4f67034a47432d51064cfaea35856b/src/generic_ui/scripts/ui.ts#L372)):</sup>
Is this to block failures which time out after a new attempt succeeds?

---

<sup>**[src/generic_ui/scripts/ui.ts, line 588 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1836#-JxFrK7Jtg7JX8WtAAEO)** ([raw file](https://github.com/uproxy/uproxy/blob/a45ff8c47a4f67034a47432d51064cfaea35856b/src/generic_ui/scripts/ui.ts#L588)):</sup>
If we keep this, I feel like this should be this.stopGettingFromInstance(data.instanceId); and in the else if below there should be this.stopGettingFromInstance(this.instanceGettingAccessFrom_);

Longer, optional suggestion: since .browserApi.stopUsingProxy was moved out and this function to only called after we have already terminated the connection (by calling stopGettingFromInstance) or if we received a disconnect message from the core, I think this function could be changed to "stoppedGetting". This would require one more change in root.ts's revertProxySettings, which would need to set disconnectedWhileProxying to false and update the icon (which is private right now though). I feel like this is better than having revertProxySettings call stopGettingInUiAndConfig when the "AndConfig" part is now handled by stopUsingProxy, and the "stopGettingInUi" part has already happened (i.e. getting status is already gone and icon is now an error icon). (Hope this made sense)

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1836)
<!-- Sent from Reviewable.io -->

bemasc commented:
Review status: 0 of 7 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

<sup>**[src/generic_ui/polymer/root.html, line 383 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1836#-JxG1riZXsgAGapszMK8)** ([raw file](https://github.com/uproxy/uproxy/blob/a45ff8c47a4f67034a47432d51064cfaea35856b/src/generic_ui/polymer/root.html#L383)):</sup>
Done.  (See updated screenshots.)

---

<sup>**[src/generic_ui/scripts/ui.ts, line 372 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1836#-JxG1cj6XgmLgQpoH1ZJ)** ([raw file](https://github.com/uproxy/uproxy/blob/a45ff8c47a4f67034a47432d51064cfaea35856b/src/generic_ui/scripts/ui.ts#L372)):</sup>
I added a comment to try to explain what this is for.

---

<sup>**[src/generic_ui/scripts/ui.ts, line 589 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1836#-JxFrK7Jtg7JX8WtAAEO)** ([raw file](https://github.com/uproxy/uproxy/blob/a45ff8c47a4f67034a47432d51064cfaea35856b/src/generic_ui/scripts/ui.ts#L589)):</sup>
Done!  Instead of making `updateIcon_` public, I just made sure to call it everywhere that changes icon-relevant state.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1836)
<!-- Sent from Reviewable.io -->

lucyhe commented:
Awesome! :+1:

---

Review status: 0 of 7 files reviewed at latest revision, all discussions resolved, some commit checks broke.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1836)
<!-- Sent from Reviewable.io -->

salomegeo commented:
Review status: 0 of 7 files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---

<sup>**[src/generic_ui/scripts/ui.ts, line 593 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1836#-JxHBFsXxJsx12I6_z0K)** ([raw file](https://github.com/uproxy/uproxy/blob/69d6ecd5e3b5f48784cc8da996b47a1279ea2d58/src/generic_ui/scripts/ui.ts#L593)):</sup>
Why did you remove this? Was it redundant?

---

<sup>**[src/generic_ui/scripts/ui.ts, line 597 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1836#-JxHEWWdGM6RrM7VfvU5)** ([raw file](https://github.com/uproxy/uproxy/blob/59ffed7a6bab6e0c40005e51e25720bf2068acf5/src/generic_ui/scripts/ui.ts#L597)):</sup>
Is there a case when this is not set?

---

<sup>**[src/generic_ui/scripts/ui.ts, line 617 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1836#-JxHEgKzuI5axURqy3eA)** ([raw file](https://github.com/uproxy/uproxy/blob/59ffed7a6bab6e0c40005e51e25720bf2068acf5/src/generic_ui/scripts/ui.ts#L617)):</sup>
I don't really understand why you need this, but in case you do, this will be false if you close and open a popup window.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1836)
<!-- Sent from Reviewable.io -->

bemasc commented:
Review status: 0 of 8 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

<sup>**[src/generic_ui/scripts/ui.ts, line 593 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1836#-JxHBFsXxJsx12I6_z0K)** ([raw file](https://github.com/uproxy/uproxy/blob/69d6ecd5e3b5f48784cc8da996b47a1279ea2d58/src/generic_ui/scripts/ui.ts#L593)):</sup>
Yes.  `disconnectedWhileProxying` is set to `true` here if there is an error, and then is set to `false` if the connection succeeds or is canceled.

---

<sup>**[src/generic_ui/scripts/ui.ts, line 597 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1836#-JxHEWWdGM6RrM7VfvU5)** ([raw file](https://github.com/uproxy/uproxy/blob/59ffed7a6bab6e0c40005e51e25720bf2068acf5/src/generic_ui/scripts/ui.ts#L597)):</sup>
I think you're right; this was redundant.  Removed.

---

<sup>**[src/generic_ui/scripts/ui.ts, line 617 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1836#-JxHEgKzuI5axURqy3eA)** ([raw file](https://github.com/uproxy/uproxy/blob/59ffed7a6bab6e0c40005e51e25720bf2068acf5/src/generic_ui/scripts/ui.ts#L617)):</sup>
Good point!  I've removed `proxySet_` entirely, and also stopped relying on `instanceGettingAccessFrom_`, which is cleared when the window is closed and opened.  Instead, I've changed `core.disconnectedWhileProxying` to include the ID of the instance from which we were disconnected.  This also required fixing the state synchronization for `instanceTryingToGetAccessFrom`, which was getting lost across window close/open.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1836)
<!-- Sent from Reviewable.io -->

salomegeo commented:
:+1: 

---

Review status: 0 of 8 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1836)
<!-- Sent from Reviewable.io -->
